### PR TITLE
Avoid calling superclass matcher where possible

### PR DESCRIPTION
**Pull request by mar-kolya**
Saves about 300ms of startup time on test app

### CODE REVIEW

File: dd-java-agent/instrumentation/hystrix-1.4/src/main/java/datadog/trace/instrumentation/hystrix/HystrixInstrumentation.java [hunk 1]
```diff
@@ -29,9 +30,11 @@ public HystrixInstrumentation() {
...
   @Override
   public ElementMatcher<TypeDescription> typeMatcher() {
-    return safeExtendsClass(
-        named("com.netflix.hystrix.HystrixCommand")
-            .or(named("com.netflix.hystrix.HystrixObservableCommand")));
+    return nameStartsWith("com.netflix.hystrix.")
+        .and(
+            safeExtendsClass(
+                named("com.netflix.hystrix.HystrixCommand")
+                    .or(named("com.netflix.hystrix.HystrixObservableCommand"))));
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/hystrix-1.4/src/main/java/datadog/trace/instrumentation/hystrix/HystrixInstrumentation.java]: I think some of these are safe, but many aren't.  For example, `HystrixCommand` is intended to be extended, so most usages of this won't match the package prefix.  I'm guessing many of these others have a similar issue.

File: dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java [hunk 2]
```diff
@@ -31,7 +32,9 @@
 
   @Override
   public ElementMatcher<TypeDescription> typeMatcher() {
-    return not(isInterface()).and(safeHasInterface(named("org.hibernate.Criteria")));
+    return nameStartsWith("org.hibernate.")
```
Comment (devinsba) [inline dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java]: I don't think this one can be done. I've seen custom `Criteria` classes before

Comment (mar-kolya) [reply dd-java-agent/instrumentation/hibernate/core-3.3/src/main/java/datadog/trace/instrumentation/hibernate/core/v3_3/CriteriaInstrumentation.java]: reverged whole hibernate and a few others

**Comment by dougqh** [comment]
The change seems like a nice improvement overall, but I'm wondering if we need to adjust our level of abstraction at some point -- not for this PR.

Specifically, I'm wondering if we should have a method on the Instrumenter that describes whether to search just a subset of packages.

I also think we could stand to create some helper methods for common idioms like...
not(isInterface()).and(hasSuper(...))

**Pull request merged**